### PR TITLE
feat(tir): add canonical parser printer and runtime coverage

### PR DESCRIPTION
**Pull request by zhen8838**
## Why
- Authored TIR could not be serialized and imported again: `as_script` only handled HIR, and modules containing `PrimFunction` raised instead of producing source.
- TIR specialization was flattened into a lowering-only `DispatchCall` statement, while half-open specialization ranges made adjacent authored variants easy to overlap.

## What
- Add visitor-based HIR/TIR printers with registry dispatch, shared Python value rendering, context-owned imports and mesh state, and source equality against five authored TIR fixtures.
- Reuse one loop header across dialects; accept scalar expressions in authored TIR `for` bounds; add authored `while`; model `Abort` as a registered effect op whose CUDA emitter traps.
- Preserve specialization structure on `PrimFunction`, author variants with `@f.specialize(...)`, and lower specialized calls directly to nested `If` / `Evaluate` / `Abort` nodes. Remove `DispatchCall`, its emitter, verifier and printer handling, and specification surface.
- Keep five H200 correctness cases beside sync, async copy, MMA, RMSNorm, and authored control flow. Fix RMSNorm runtime argument localization to match sibling CUDA operations.
- Final verification at `f8777af`: source `tir-parser-final-source-13` (942 tests, 0 failures/errors) and installed `tir-parser-final-installed-7` (161 tests, 0 failures/errors).

## Contract
- `as_script` accepts `PrimFunction` and mixed HIR/TIR modules; authored TIR is canonical when printed source equals its fixture.
- `DimVarRangePat(lo, hi)` is closed `[lo, hi]`; the `DimVar(lo, hi)` envelope remains half-open `[lo, hi)`. Named maximum and capacity constants are valid specialization values, and envelopes use one-past-the-maximum bounds.
- The TIR `Stmt` tree is no longer frozen and uses `unsafe_hash=True`. `PrimFunction` carries `specializations`, `variants`, `_sealed`, and `_display_name`.
- Specialized calls lower structurally to nested `If` / `Evaluate` / `Abort`; no `DispatchCall` IR node or authored surface remains.
- Regenerate the parser and public IR specifications. The `placed-layout` to `placed-shape` correction repairs pre-existing drift introduced by #150 before #154; this branch did not introduce it.

## Risk
- HIR printer output currently has no regression test after removal of its generated corpus snapshot; the TIR golden fixtures do not cover HIR output drift.
- `while` has a CUDA emitter but no GPU end-to-end case. Dynamic CTA remains out of scope.
- Hopper WGMMA/TMA expansion is deferred to a separate plan under D11.

### CODE REVIEW

File: src/tilefoundry/inspection/_python_render.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+from tilefoundry.ir.core.pattern import DimVarRangePat, Pattern
+from tilefoundry.ir.types import DType, TensorType
+from tilefoundry.ir.types.shard.layout import ComposedLayout, Layout, LayoutBase
+from tilefoundry.ir.types.shard.mesh import Mesh
+from tilefoundry.ir.types.shard.shard_layout import Broadcast, Partial, ShardLayout, Split
+from tilefoundry.ir.types.storage import StorageKind
+
+
+def dtype_str(dtype: DType) -> str:
```
Comment (zhen8838) [inline src/tilefoundry/inspection/_python_render.py]: These functions should have been written many times. Please use them again.

File: src/tilefoundry/inspection/tir_printer.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            elif isinstance(value, (str, int, float, bool, tuple)):
+                value = repr(value)
+            else:
+                raise NotImplementedError(f"TIR printer has no canonical attribute form for {type(value).__name__}")
+            args.append(f"{p.name}={value}")
+        return f"T.{op_name}({', '.join(args)})"
+    raise NotImplementedError(f"TIR printer has no canonical expression form for {type(expr).__name__}")
+
+
+def _emit_stmt(stmt, indent: str, lines: list[str], mesh_bindings=None) -> None:
```
Comment (zhen8838) [inline src/tilefoundry/inspection/tir_printer.py]: It should be to write an stmt visitor to do this, and then it can reflect on each op's own printer implementation.

File: src/tilefoundry/parser/pattern_nodes.py [hunk 3]
```diff
@@ -4049,6 +4088,165 @@ def construct(match, children, context):
...
+    @staticmethod
+    def construct(match, children, context):
+        frame = children["header"]
+        context.lexical_scope.pop_frame()
+        return runtime.For(frame.induction_var, frame.start, frame.stop, frame.step, children["body"])
+
+    RULES: ClassVar[tuple[AstRule[Any], ...]] = ()
+
+
+class TirIfPattern(ElementPattern):
```
Comment (zhen8838) [inline src/tilefoundry/parser/pattern_nodes.py]: Why do for and if need to add patterns separately and can't be reused?

File: tests/fixtures/tir/async_sync.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    def async_stage_device(
+        a: Tensor[(128, 4), "f32"], b: Tensor[(128, 4), "f32"]
+    ):
+        with Mesh((Topology("thread", 128),), Layout((128,), (1,)), ("t",)) as m:
+            a_view = T.tensor_view(
+                a,
+                layout=ShardLayout(Layout((128, 4), (4, 1)), (S(0),), m),
+            )
+            shared = T.alloc_tensor(
+                TensorType(
```
Comment (zhen8838) [inline tests/fixtures/tir/async_sync.py]: If the input of T.alloc_tensor is TensorType, it should be able to be parsed in the form of Tensor[].

File: tests/inspection/test_tir_roundtrip.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+from tilefoundry.ir.tir.dispatch import DispatchCall
+from tilefoundry.ir.tir.launch import Launch
+from tilefoundry.ir.tir.shape import ShapeOf
+from tilefoundry.ir.tir.stmts import Abort, For, Sequential
+from tilefoundry.ir.types import DType, TensorType, UnitType
+from tilefoundry.parser.ast_pattern import ParseError
+from tilefoundry.target import CpuTarget
+from tilefoundry.visitor_registry import typeinfer_registry
+
+TIR_SUPPORT_MATRIX = {
```
Comment (zhen8838) [inline tests/inspection/test_tir_roundtrip.py]: There shouldn't be these dictionaries.

File: tests/inspection/test_tir_roundtrip.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    "square": (TirSquare, "TirSquare"),
+    "rmsnorm": (TirRmsnorm, "TirRmsnorm"),
+    "mma": (MmHandwritten, "MmHandwritten"),
+    "sync": (SyncSquare, "SyncSquare"),
+    "async_copy": (AsyncStage, "AsyncStage"),
+    "remaining_ops": (current_tir_ops, "current_tir_ops"),
+}
+
+
+def _assert_lint_clean(source: str) -> None:
```
Comment (zhen8838) [inline tests/inspection/test_tir_roundtrip.py]: What is this for?

File: tests/inspection/test_tir_roundtrip.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            result_type = result_type.rsplit(".", 1)[-1]
+        inferred_forms[name] = (
+            "effect-form" if result_type in (UnitType, "UnitType") else "value-form"
+            if result_type in (TensorType, "TensorType")
+            else None
+        )
+        assert inferred_forms[name] == expected
+
+
+def test_unregistered_printable_tir_nodes_are_named() -> None:
```
Comment (zhen8838) [inline tests/inspection/test_tir_roundtrip.py]: These tests are all wrong. It's very simple as_script(xxx module) == module source, module source should all come from tests fixtrue

File: tests/integration/test_tir_perf.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import json
+import math
+import statistics
+from dataclasses import asdict, dataclass
+
+import pytest
+import torch
+
+import tilefoundry
+from tests.integration.test_tir_runtime import (
```
Comment (zhen8838) [inline tests/integration/test_tir_perf.py]: Don't put this test in the integration, and delete all the relevant content.

File: tests/integration/test_tir_runtime.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+TIR_RUNTIME_CASES = (
+    TirRuntimeCase("square", TirSquare, _square_invocation),
+    TirRuntimeCase("rmsnorm", TirRmsnorm, _rmsnorm_invocation),
+    TirRuntimeCase("mma", MmHandwritten, _mma_invocation),
+    TirRuntimeCase("async_copy", AsyncStage, _async_copy_invocation),
+    TirRuntimeCase("sync", SyncSquare, _sync_invocation),
+)
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="requires CUDA")
```
Comment (zhen8838) [inline tests/integration/test_tir_runtime.py]: Delete the same

File: src/tilefoundry/inspection/_python_render.py (hunk 1 above)
Comment (zhen8838) [reply src/tilefoundry/inspection/_python_render.py]: Resolved in a5b96d4: shared rendering now lives in PythonPrinter; _python_render.py was removed.

File: src/tilefoundry/inspection/tir_printer.py (hunk 2 above)
Comment (zhen8838) [reply src/tilefoundry/inspection/tir_printer.py]: Resolved in a5b96d4: TirPrinter is a StmtVisitor with visit_* methods and register_tir_printer dispatch.

File: src/tilefoundry/parser/pattern_nodes.py (hunk 3 above)
Comment (zhen8838) [reply src/tilefoundry/parser/pattern_nodes.py]: Resolved in eeca8b5: HIR and TIR share ForPattern/IfPattern and dispatch construction by dialect.

File: tests/fixtures/tir/async_sync.py (hunk 4 above)
Comment (zhen8838) [reply tests/fixtures/tir/async_sync.py]: Resolved in eeca8b5: committed TIR fixtures use Tensor[...] annotations for alloc_tensor.

File: tests/inspection/test_tir_roundtrip.py (hunk 5 above)
Comment (zhen8838) [reply tests/inspection/test_tir_roundtrip.py]: Resolved in eeca8b5: support matrices and schema-form dictionaries were removed in favor of fixture golden equality.

File: tests/inspection/test_tir_roundtrip.py (hunk 6 above)
Comment (zhen8838) [reply tests/inspection/test_tir_roundtrip.py]: Resolved in eeca8b5: the per-test ruff subprocess helper was removed; committed fixtures are linted normally.

File: tests/inspection/test_tir_roundtrip.py (hunk 7 above)
Comment (zhen8838) [reply tests/inspection/test_tir_roundtrip.py]: Resolved through eeca8b5 and b1f61f5: as_script(module) is compared byte-for-byte with committed fixture source; HIR has a complete checked-in golden.

File: tests/integration/test_tir_perf.py (hunk 8 above)
Comment (zhen8838) [reply tests/integration/test_tir_perf.py]: Resolved in eeca8b5: test_tir_perf.py and its timing infrastructure were deleted.

File: tests/integration/test_tir_runtime.py (hunk 9 above)
Comment (zhen8838) [reply tests/integration/test_tir_runtime.py]: Resolved in eeca8b5: test_tir_runtime.py was deleted and each direct GPU oracle moved beside its tested op; control flow has its own integration test.

File: docs/spec/parser.md [hunk 10]
```diff
@@ -156,6 +160,8 @@ type-annotation       ::= tensor
                           | scalar-type
 signature             ::= (name ':' type-annotation (',' name ':' type-annotation)*)?
 return-type           ::= type-annotation
+tir-if                ::= if cond-node block (block)?
```
Comment (zhen8838) [inline docs/spec/parser.md]: Why are there still two independent grammars, if of TIR and while of TIR?

File: docs/spec/tir.md [hunk 11]
```diff
@@ -356,6 +356,21 @@ class DispatchCall(Stmt):
 ([§1.4](#14-evaluate)); each `case_calls[i]` is an
 `Evaluate(SymbolRef, args)` invoking that case's specialized callee.
 
+The inspection printer uses a statement suite for the fallback, so every
+fallback `Stmt` remains readable rather than being forced into an expression
+or lambda:
+
+```python
```
Comment (zhen8838) [inline docs/spec/tir.md]: Why is this not an OP?

File: pyproject.toml [hunk 12]
```diff
@@ -119,5 +119,6 @@ ignore = ["E501", "F821", "E741"]
 # authoring-surface ergonomics. Internal implementation code in
 # `src/` must NOT use star imports (no exception here).
 "tests/**/*.py" = ["F403", "F405"]
+"tests/fixtures/logical/canonical_*.py" = ["F401", "I001"]
```
Comment (zhen8838) [inline pyproject.toml]: Why do you want to be exempted here?

File: src/tilefoundry/ir/types/shard/mesh.py [hunk 13]
```diff
@@ -17,6 +18,9 @@ class Topology:
 
     size: "ShapeDim"
 
+    def to_python(self) -> PythonExpr:
```
Comment (zhen8838) [inline src/tilefoundry/ir/types/shard/mesh.py]: Is there a better way to write here? It feels a little complicated to rely on it like this.

File: docs/spec/parser.md (hunk 10 above)
Comment (zhen8838) [reply docs/spec/parser.md]: Resolved in eb18619 and subsequent M8 cleanup: HIR and TIR now share one loop-header grammar, while dialect-specific construction is selected by parser rules. The generated grammar uses `if` and `while` without `tir_` prefixes.

File: docs/spec/tir.md (hunk 11 above)
Comment (zhen8838) [reply docs/spec/tir.md]: Resolved through 0ad87cf and M12: authored abort is a registered effect-form `T` Op. The lowering-only `DispatchCall` node, emitter, verifier, printer form, and spec section were removed; specialization calls now lower structurally to `If` / `Evaluate` / `Abort`.

File: pyproject.toml (hunk 12 above)
Comment (zhen8838) [reply pyproject.toml]: Resolved in M11: the HIR-only canonical fixtures were removed together with this per-file lint exemption. TIR golden fixtures remain ordinary committed, linted source files.

File: src/tilefoundry/ir/types/shard/mesh.py (hunk 13 above)
Comment (zhen8838) [reply src/tilefoundry/ir/types/shard/mesh.py]: Resolved in 2a79940: Python-source rendering was removed from IR types and centralized in `PythonPrinter`; IR objects remain independent of inspection. Target rendering continues to use each Target to_python result through the print context.

File: tests/fixtures/placed/gqa_decode.py [hunk 14]
```diff
@@ -81,7 +81,7 @@ def gqa_online_attend(
 
         pass
 
-    @gqa_online_attend.specialize(DimVarRangePat("ctx_len", 0, SMALL_CONTEXT_T))
+    @gqa_online_attend.specialize(DimVarRangePat("ctx_len", 0, SMALL_CONTEXT_T - 1))
```
Comment (zhen8838) [inline tests/fixtures/placed/gqa_decode.py]: Why should the upper limit be reduced by one? First of all, what we now describe in dim var is actually the value range of shape. If its upper limit is 16K, then it can get 16K instead of 16K-1.

Comment (zhen8838) [reply tests/fixtures/placed/gqa_decode.py]: Resolved in cae7c00 and 23427cd. `SMALL_CONTEXT_T=4096` and `MAX_CTX=262144` are now valid inclusive maxima: the closed specialization arms are `[0, SMALL_CONTEXT_T]` and `[SMALL_CONTEXT_T + 1, MAX_CTX]`, while the half-open DimVar envelope is `[0, MAX_CTX + 1)`. Boundary-selection and out-of-envelope tests were updated, and CLI output now renders specialization ranges as closed `[lo, hi]`. Full verification at 23427cd: source 942/0/0/10 (`tir-parser-final-source-12`), installed 161/0/0/0 (`tir-parser-final-installed-6`).
